fix: support objects for function_call in chat completion

ryanskidmore opened the pull request:
This PR is a duplicate of #374, and addresses issue #378. I've copied a majority of the changes from @j178's PR (thank you) and addressed the feedback that was raised in that PR. I've only applied this to the chat completion endpoint as that's what I needed these changes for. Unfortunately there was already a struct named `FunctionCall` so unfortunately had to go with the rather ugly naming of `ChatCompletionRequestFunctionCall`. Happy to hear suggestions for better names!

Happy to close this in favour of @j178's PR if they wish given most of this is their code. I needed this for a project I'm working on, otherwise I would have waited on the other PR.

jmacwhyte commented:
This is incorrect. There are two different properties called "function_call" in the API:

1. When making a request. ChatCompletionRequest.FunctionCall is what is being discussed in the issues you mentioned, and can be either "none", "auto", or an object with a "name" property specifying which function you want to force the AI to call.
2. When the AI responds to a request, and decides to call a function you have defined. In this instance, ChatCompletionMessage.FunctionCall will contain an object of the following format:

`{
Name: [function name],
Arguments: {[argument name]: [argument value], ...},
}
`

Your PR appears to treat number 2 as if it is number 1.

ryanskidmore commented:
> This is incorrect. There are two different properties called "function_call" in the API:
> 
> 1. When making a request. ChatCompletionRequest.FunctionCall is what is being discussed in the issues you mentioned, and can be either "none", "auto", or an object with a "name" property specifying which function you want to force the AI to call.
> 2. When the AI responds to a request, and decides to call a function you have defined. In this instance, FinishReason will be set to `function_call` and ChatCompletionMessage.FunctionCall will contain an object of the following format:
> 
> ```
> {
> Name: [function name],
> Arguments: {[argument name]: [argument value], ...},
> }
> ```
> 
> Your PR appears to treat number 2 as if it is number 1.

@jmacwhyte I think you may have misread the code. This PR adds a `FunctionCall` field to the `ChatCompletionRequest` struct - addressing (1). This PR does not make any changes related to (2), which is already functional.

jmacwhyte commented:
> @jmacwhyte I think you may have misread the code. This PR adds a `FunctionCall` field to the `ChatCompletionRequest` struct - addressing (1). This PR does not make any changes related to (2), which is already functional.

Crap, you're right. It looked like the FunctionCall you were changing was the one inside of ChatCompletionMessage, because the code in between was hidden. My mistake!

jmacwhyte commented:
I've look at it again (correctly, this time) and I think this is a very good solution. Nice job!

ryanskidmore commented:
> I've look at it again (correctly, this time) and I think this is a very good solution. Nice job!

Thanks! I can't take much (or really, any) credit for it though 😅

sashabaranov commented on chat.go:
```diff
@@ -55,7 +56,26 @@ type ChatCompletionRequest struct {
+	FunctionCall any `json:"function_call,omitempty"`
+}
+
+func checkFunctionCall(value any) bool {
```
Hey, thank you for the PR! The name of this function sounds a bit too vague. Maybe let's go with something a bit more specific like `checkIsChatCompletionFunctionCall`?

ryanskidmore replied:
@sashabaranov Good idea - I've just pushed a change to rename the function to exactly that.

steeve commented:
I have another take at this, that is type safe and without reflection if you folks want:

```go
var (
	FunctionCallNone = &FunctionCall{}
	FunctionCallAuto = &FunctionCall{}
)

func (f *FunctionCall) MarshalJSON() ([]byte, error) {
	switch f {
	case FunctionCallNone:
		return []byte(`"none"`), nil
	case FunctionCallAuto:
		return []byte(`"auto"`), nil
	default:
	}
	type functionCallAlias FunctionCall
	return json.Marshal((*functionCallAlias)(f))
}
```

```go
	FunctionCall     *FunctionCall           `json:"function_call,omitempty"`
```

tartavull commented:
What's missing to get this merged? :)

sashabaranov commented:
Okay, we have a conflict here now based on work done in #377 and #407

@ryanskidmore I love the additional testing and type-checking. Would it be possible to resolve conflicts with the latest updates? Thanks!

jmacwhyte commented on chat.go:
```diff
@@ -55,7 +56,26 @@ type ChatCompletionRequest struct {
+	FunctionCall any `json:"function_call,omitempty"`
+}
+
+func checkIsChatCompletionFunctionCall(value any) bool {
```
Should this maybe be named something like `validateChatCompletionFunctionCall`? The current name makes me think a return of `true` would mean it is a function call, but `true` means it is a valid value (even if there is no function call involved).

jmacwhyte commented:
I also think @steeve 's solution is very clever (and compact), how do we decide what is more maintainable?

ryanskidmore commented:
@sashabaranov (cc @jmacwhyte @tartavull @steeve & any others interested) I've rebased my fork onto master and resolved the outstanding conflicts. With the recent changes that were made in other PRs I took a bit of a different direction which I appreciate may conflict with intentional decisions made in other PRs - so please feel free to give me feedback on those directional changes.

The `FunctionCall` field moving to be an `any` type made the usage very non-intuitive and confusing (from the perspective of someone who uses this library). You have to read the documentation to understand what to put in the `FunctionCall` field which IMO isn't great from a design perspective. I've moved this field back to being a pointer to the `FunctionCall` struct and adopted @steeve's novel approach which removes the need for this field to be an `any` type. I've changed the validation to match this change accordingly which ensures the field is populated with a valid value.

jmacwhyte commented on chat.go:
```diff
@@ -34,12 +36,37 @@ type ChatCompletionMessage struct {
+}
+
+func validateChatCompletionFunctionCall(v *FunctionCall) bool {
+	if v == nil || v == FunctionCallNone || v == FunctionCallAuto || v.Name != "" {
```
Now that this is so simple (one line), instead of making it a separate function could you just put it inline at line number 146 below?

vvatanabe replied:
I kindly suggest we keep this as a function. The reasons being, it promotes enhances readability, facilitates testing, and simplifies refactoring. I believe these are crucial elements in preserving the quality of our code.

jmacwhyte replied:
I'm good with that!

jmacwhyte commented:
I think it all looks great!

vvatanabe commented on chat.go:
```diff
@@ -34,12 +36,37 @@ type ChatCompletionMessage struct {
+		return []byte(`"auto"`), nil
+	default:
+	}
+	type functionCallAlias FunctionCall
+	return json.Marshal((*functionCallAlias)(f))
```
Developers unfamiliar with Go may be a bit confused, so please add a comment that the use of type aliases prevents recursive calls.

arrix replied:
Learned something from this comment 👍. It would be helpful to clarify this is a type definition rather than an alias. In go's terms, an alias is an identifier bound to a type so it receives all the methods from it too.

sullrich commented:
Hoping we can keep this alive, this is needed with modern ai api calls.  Can someone review + commit?

jmacwhyte commented:
I just assumed this was accepted 6 months ago, and haven't looked at it much since! @sashabaranov can you update us on the latest?

ryanskidmore commented:
I'm going to close this for now given the number of changes that have happened in this code over the last ~6 months. These fields have been deprecated now so I don't see a need to revive this.
